node-wasm/src/utils.rs

```diff
@@ -73,6 +79,134 @@ pub(crate) fn js_value_from_display<D: fmt::Display>(value: D) -> JsValue {
+    fn to_error<C>(self, context: C) -> Result<T, JsError>
+    where
+        C: fmt::Display + Send + Sync + 'static,
+    {
+        self.map_err(|e| {
+            let error_str = match e.dyn_ref::<JsString>() {
+                Some(s) => format!("{context}: {s}"),
+                None => format!("{context}"),
+            };
+            JsError::new(&error_str)
+        })
+    }
+}
+
+pub(crate) trait WorkerSelf {
+    type GlobalScope;
+
+    fn worker_self() -> Self::GlobalScope;
+    fn is_worker_type() -> bool;
+}
+
+impl WorkerSelf for SharedWorker {
+    type GlobalScope = SharedWorkerGlobalScope;
+
+    fn worker_self() -> Self::GlobalScope {
+        JsValue::from(js_sys::global()).into()
+    }
+
+    fn is_worker_type() -> bool {
+        js_sys::global().has_type::<Self::GlobalScope>()
+    }
+}
+
+impl WorkerSelf for Worker {
+    type GlobalScope = DedicatedWorkerGlobalScope;
+
+    fn worker_self() -> Self::GlobalScope {
+        JsValue::from(js_sys::global()).into()
+    }
+
+    fn is_worker_type() -> bool {
+        js_sys::global().has_type::<Self::GlobalScope>()
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub(crate) enum JsResult<T, E>
+where
+    T: JsCast + Debug,
+    E: Debug,
+{
+    #[serde(with = "serde_wasm_bindgen::preserve")]
+    Ok(T),
+    Err(E),
+}
+
+// once try_trait_v2 is stabilised, this can go
+impl<T, E> JsResult<T, E>
+where
+    T: JsCast + Debug,
+    E: Serialize + DeserializeOwned + Debug,
+{
+    pub fn into_result(self) -> Result<T, E> {
+        self.into()
+    }
+}
+
+impl<T, E> From<Result<T, E>> for JsResult<T, E>
+where
+    T: JsCast + Debug,
+    E: Serialize + DeserializeOwned + Debug,
+{
+    fn from(result: Result<T, E>) -> Self {
+        match result {
+            Ok(v) => JsResult::Ok(v),
+            Err(e) => JsResult::Err(e),
+        }
+    }
+}
+
+impl<T, E> From<JsResult<T, E>> for Result<T, E>
+where
+    T: JsCast + Debug,
+    E: Serialize + DeserializeOwned + Debug,
+{
+    fn from(result: JsResult<T, E>) -> Self {
+        match result {
+            JsResult::Ok(v) => Ok(v),
+            JsResult::Err(e) => Err(e),
+        }
+    }
+}
+
+const CHROME_USER_AGENT_DETECTION_STR: &str = "Chrome/";
+
+// currently there's issue with SharedWorkers on Chrome, where restarting lumina's worker
+// causes all network connections to fail. Until that's resolved detect chrome and apply
+// a workaround.
+pub(crate) fn is_chrome() -> bool {
+    let mut user_agent = None;
+    if let Some(window) = window() {
+        user_agent = Some(window.navigator().user_agent());
+    };
+    if let Some(worker_scope) = JsValue::from(js_sys::global()).dyn_ref::<WorkerGlobalScope>() {
+        user_agent = Some(worker_scope.navigator().user_agent());
+    }
+
+    if let Some(user_agent) = user_agent {
+        user_agent
+            .as_deref()
+            .unwrap_or("")
+            .contains(CHROME_USER_AGENT_DETECTION_STR)
+    } else {
+        false
+    }
+}
```

Comment:
It would be nice if the logic runs once and then return a cached result

Reply:
True, but do we actually need it? Currently we do detection once and fingers crossed this doesn't increase. 
We could bring in `memoize` crate or do caching manually when/if this is called multiple times.